Sources/CacheAdvance/CacheReader.swift

```diff
@@ -0,0 +1,130 @@
+//
+//  Created by Dan Federman on 12/26/19.
+//  Copyright © 2019 Dan Federman.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+
+final class CacheReader {
+
+    // MARK: Initialization
+
+    /// Creates a new instance of the receiver.
+    ///
+    /// - Parameters:
+    ///   - file: The file URL indicating the desired location of the on-disk store. This file should already exist.
+    ///   - overwriteOldMessages: When `true`,  the cache encodes a pointer to the oldest message after the newest message marker.
+    init(forReadingFrom file: URL, overwriteOldMessages: Bool) throws {
+        reader = try FileHandle(forReadingFrom: file)
+        self.overwriteOldMessages = overwriteOldMessages
+    }
+
+    deinit {
+        try? reader.closeHandle()
+    }
+
+    // MARK: Internal
+
+    var offsetInFileOfOldestMessage: UInt64 = 0
+    var offsetInFileAtEndOfNewestMessage: UInt64 = 0
+
+    var offsetInFile: UInt64 {
+        reader.offsetInFile
+    }
+
+    /// Returns the next encodable message, seeking to the beginning of the next message.
+    func nextEncodedMessage() throws -> Data? {
+        let startingOffset = offsetInFile
+
+        guard startingOffset != offsetInFileAtEndOfNewestMessage else {
+            // We're at the last message.
+            return nil
+        }
+
+        switch try nextEncodedMessageSpan() {
+        case let .span(messageLength):
+            let message = try reader.readDataUp(toLength: Int(messageLength))
```

Comment:
This is a really good point. I don't know how I didn't notice this before. At an absolute minimum there should be a clarifying comment here.

Most file handle APIs use `UInt64`. Only this one uses `Int`, which is absolutely surprising.

> If a `FileHandle` can only read data whose length is less than or equal to the length of an `Int`, should we make `MessageSpan` be an `Int` instead of an `Int32`?

We can't do this, since we want our files to be capable of being read from a different device than the device that wrote them (which is why we bothered with standardizing endianness). If we relied on `Int`, and then we wrote a cache on a 32bit (`Int32`) machine and tried to read the cache from a 64bit (`Int64`) machine, we wouldn't be able to do so.

That said, the above scenario is only a problem if we support 32bit machines. So maybe this example would make more sense in a world where 128bit machines have been introduced. In any case, the rule here is: any information we store to disk must have a fixed size. Once the size of information on disk is variable, it becomes extremely easy to corrupt a file (or interpret a file as corrupt)

> If there are circumstances where an `Int32` can be longer than an `Int` and we actually had messages that were longer than an `Int`, it seems like we wouldn’t be able to read them.

You're right. Not with the current code at least. The maximum message length must be within `Int` on the local machine. Which means that our largest supported message is really only 2GB.

Your comment made me dig in, and, well, this isn't the only place we're limiting the possible size of a message to the `Int` space. `Data`'s `count` returns an `Int`, so trying to write a message larger than 2GB on a 32bit machine might crash.

It would probably be good for us to either:
1. Decide we only support 64bit machines. Given the operating systems we support, that seems like a reasonable restriction.
1. Enforce that a message is smaller than Int32.max.

Reply:
Just remembered that macOS 10.14 still supports 32bit devices. Sigh.

I think it's fair to say that using this cache for messages that are larger than 2GB is user error. I'll add something to the README.

Reply:
Added a `throw` in 82893b297747c879c028487d64609c8e9fa6adce

Reply:
I forgot that `Int` changed sizes depending on the machine.
This all sounds good 👍 
Thanks for digging in.